File: #1857.largest-color-value-in-a-directed-graph.cpp
class Solution {
public:
    int largestPathValue(string colors, vector<vector<int>>& edges) {

        int size = colors.size();
        vector<vector<int>> adj(size);
        vector<int> indegree(size);
        for (auto& edge : edges) {
            adj[edge[0]].push_back(edge[1]);
            indegree[edge[1]]++;
        }

        vector<vector<int>> count(size, vector<int>(26));
        queue<int> q;
        for (int i = 0; i < size; i++) {
            if (indegree[i] == 0) {
                q.push(i);
            }
        }

        int result = 0;
        int seen = 0;
        while(!q.empty()){
            int node = q.front();
            q.pop();
            result = std::max(result,
                ++count[node][colors[node] - 'a']);
            seen++;
            for (auto& neighbor : adj[node]) {
                for (int i = 0; i < 26; i++) {
                    count[neighbor][i] =
                        std::max(count[neighbor][i], count[node][i]);
                }

                indegree[neighbor]--;
                if (indegree[neighbor] == 0) {
                    q.push(neighbor);
                }
            }
        }
        return seen == size ? result : -1;
    }
};

File: #1857.largest-color-value-in-a-directed-graph.js
/**
 * @param {string} colors
 * @param {number[][]} edges
 * @return {number}
 */
var largestPathValue = function(colors, edges) {
    const n = colors.length;
    let adj = {};
    let inDeg = new Array(n).fill(0); 
    for (let [u, v] of edges) {
        if (!adj[u]) {
            adj[u] = [];
        }
        adj[u].push(v);
        inDeg[v]++;
    }

    let q = [];
    for (let i = 0; i < n; i++) {
        if (inDeg[i] === 0) {
            q.push(i);
        }
    }

    if (q.length == 0) {
        return -1;
    }

    let nodesProcessed = 0, maxColor = 0;
    let nodeColor = new Array(n).fill().map(() => new Array(26).fill(0));

    while (q.length > 0) {
        let node = q.shift();
        nodesProcessed++;
        let colorIndex = colors[node].charCodeAt(0) - 'a'.charCodeAt(0);
        nodeColor[node][colorIndex]++;
        maxColor = Math.max(maxColor, nodeColor[node][colorIndex])

        for (let neighbor of (adj[node] || [])) {
            inDeg[neighbor]--;
            if (inDeg[neighbor] === 0) {
                q.push(neighbor);
            }

            for (let i = 0; i < 26; i++) {
                nodeColor[neighbor][i] = Math.max(nodeColor[neighbor][i], nodeColor[node][i]);
            }
        }
    }

    return nodesProcessed === n ? maxColor : -1; 
};
